npm/pkg/dataplane/ipsets/ipsetmanager.go

```diff
@@ -23,261 +25,346 @@ func (iMgr *IPSetManager) exists(name string) bool {
+	}
+
+	list := iMgr.setMap[listName]
+	if list.Kind != ListSet {
+		return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s is not a list set", listName))
+	}
+
+	for _, memberName := range memberNames {
+		if listName == memberName {
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s cannot be added to itself", listName))
+		}
+		if !iMgr.exists(memberName) {
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s does not exist", memberName))
 		}
+		member := iMgr.setMap[memberName]
 
-		// check if Set is a member of List
-		_, exists = list.MemberIPSets[setName]
-		if !exists {
-			return nil
+		// Nested IPSets are only supported for windows
+		// Check if we want to actually use that support
+		if member.Kind != HashSet {
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s is not a hash set and nested list sets are not supported", memberName))
 		}
+	}
+	return nil
+}
 
-		// delete IPSet from the list
-		delete(list.MemberIPSets, setName)
-		set.DecIpsetReferCount()
-		// Update metrics of the IpSet
-		metrics.RemoveEntryFromIPSet(listName)
+func (iMgr *IPSetManager) addMemberIPSet(listName, memberName string) {
+	list := iMgr.setMap[listName]
+	member := iMgr.setMap[memberName]
+
+	list.MemberIPSets[member.Name] = member
+	member.incIPSetReferCount()
+	metrics.AddEntryToIPSet(list.Name)
+	listIsInKernel := list.shouldBeInKernel()
+	if listIsInKernel {
+		iMgr.incKernelReferCountAndModifyCache(member)
 	}
-	iMgr.updateDirtyCache(listName)
+}
 
-	return nil
+func (iMgr *IPSetManager) removeMemberIPSet(listName, memberName string) {
+	list := iMgr.setMap[listName]
+	member := iMgr.setMap[memberName]
+
+	delete(list.MemberIPSets, member.Name)
+	member.decIPSetReferCount()
+	metrics.RemoveEntryFromIPSet(list.Name)
+	listIsInKernel := list.shouldBeInKernel()
+	if listIsInKernel {
+		iMgr.decKernelReferCountAndModifyCache(member)
+	}
 }
 
 func (iMgr *IPSetManager) DeleteList(name string) error {
 	iMgr.Lock()
 	defer iMgr.Unlock()
-	set, exists := iMgr.setMap[name] // check if the Set exists
-	if !exists {
-		return npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("member ipset %s does not exist", set.Name))
+	if !iMgr.exists(name) {
+		return npmerrors.Errorf(npmerrors.DestroyIPSet, false, fmt.Sprintf("ipset %s does not exist", name))
 	}
 
-	if !set.CanBeDeleted() {
-		return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("ipset %s cannot be deleted", set.Name))
+	set := iMgr.setMap[name]
+	if !set.canBeDeleted() {
+		return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("ipset %s cannot be deleted", name))
 	}
 
+	// the set will not be in the kernel since there are no references, so there's no need to update the dirty cache
 	delete(iMgr.setMap, name)
+	metrics.DecNumIPSets()
 	return nil
 }
 
 func (iMgr *IPSetManager) DeleteSet(name string) error {
```

Comment:
I think we can have a single delete function like createIPSet we can have deleteIPSet